### ISSUE TITLE
Add Javadocs to templates

### ISSUE THREAD

**Issue by jdimeo** [eclipse/eclipse-collections#351]
Has this already come up? I'm transitioning code from Trove over to Eclipse Collections and I want to make sure the semantics are the same. Having Java docs as part of the templates that generate the specialized classes would be very helpful so I don't have to track down documentation. For example, what kind of exception is thrown by `getOrThrow()`? Thanks!

**Comment by nikhilnanivadekar** [comment]
@jdimeo this has not come up before. We generally add Javadocs to new methods and clean up any old methods when we touch them. Are there any specific methods you would like us to prioritize?
Also, we would welcome contributions if you would like to add.
Thanks for transitioning from Trove to Eclipse Collections!

**Comment by jdimeo** [comment]
I will definitely try to contribute some in the coming weeks! The top priority ones are the "bargain buy" methods - obviously get, put, add, remove type methods are pretty self-explanatory. My questions were with the ones that two or more things at once for performance. Trove has `adjustOrPutValue(key, 1, 1)` (the first being the delta and the latter being the value to put if key is absent) and you all have `addToValue(key, 1)`- I dug into the source of an impl to confirm that the 1 is also inserted if key is absent (vs. Trove's `adjustValue(key, 1)` which only adjusts if key is present).

A sampling of methods that I'm not 100% sure what they do from looking at them:
- `addToValue`
- `detectIfNone`
- `getIfAbsent(Put(With(Key)))`
- `getOrThrow`
- `injectInto`
- `removeKeyIfAbsent`

A related task that I or someone could do is write a "Trove migration guide" which has things like `adjustOrPutValue` = `addToValue`. In addition, you guys don't have the notion of a load factory right? (i.e. more dense hashing (less RAM but slower and more collisions) vs. more sparse hashing). And it looks like the default value can't be customized like it can with Trove? e.g. instead of returning `0.0` if a key is not present, return `Double.NEGATIVE_INFINITY`.

**Comment by donraab** [comment]
These all sound like great additions to me.  This would be a terrific contribution and would be very much appreciated.  I think the best thing to do in cases like you have already with addToValue is to read the implementation and the supporting tests and assume that the contract in the Javadoc should be based on the tests.  A few of the committers who review the proposed Javadoc can determine if the current tests, implementation and proposed description and contract make sense or if a something should be added to the existing tests or fixed in the implementation and then will raise an issue.

**Comment by jdimeo** [comment]
I still haven't forgotten about this and will revisit now that I'm getting my feet wet with my first PR #492

**Comment by nikhilnanivadekar** [comment]
Thanks for your help and contribution :smile:

**Comment by motlin** [comment]
Bump. I think these three methods in [primitivePrimitiveMap.stg](https://github.com/eclipse/eclipse-collections/blob/master/eclipse-collections-code-generator/src/main/resources/api/map/primitivePrimitiveMap.stg) deserve javadoc.

```java
    <type2> get(<type1> key);
    <type2> getIfAbsent(<type1> key, <type2> ifAbsent);
    <type2> getOrThrow(<type1> key);
```

**Comment by jdimeo** [comment]
@nikhilnanivadekar @donraab I finally am revisiting this- so sorry for the delay. I updated my fork to tip of `master` just now and got `mvnw` working. By the way, nothing in the README or contributing guide that I saw indicated how to get the Maven wrapper installed- I found the command in `.travis.yml`
```
mvn -N io.takari:maven:wrapper -Dmaven="3.6.3"
```

Anyway, once I got that working, I couldn't get the build to complete because of Checkstyle failures. They are all "Carriage return" fails e.g.
```
[ERROR] src/main/java/org/eclipse/collections/codegenerator/tools/package-info.java:[12] (regexp) RegexpMultiline: Carriage return.
[ERROR] src/main/java/org/eclipse/collections/codegenerator/tools/package-info.java:[13] (regexp) RegexpMultiline: Carriage return.
[ERROR] src/main/java/org/eclipse/collections/codegenerator/tools/package-info.java:[14] (regexp) RegexpMultiline: Carriage return.
```

I'm on Windows, so I popped over to WSL to run the build on Linux, and the same result happened. What am I doing wrong? surely these checks don't fail for everyone else... once I can get the source files generated I was going to see what gaps still remained for the Javadocs.

**Comment by nikhilnanivadekar** [comment]
Thanks @jdimeo for the feedback. You can skip the checkstyle in local, we all run it directly on the remote builds as it is faster that way. If you want to run it in local then I would suggest comment out the carriage return check in checkstyle.xml. Like you said it is due to the Windows vs Linux line endings. I had a fix for that in my local, I will need to dig it up and update it.
@motlin can you please add the instructions to get the maven wrapper in contributing guide? Thanks!

**Comment by jdimeo** [comment]
You learn something new every day. I hadn't seen or used `<defaultGoal>` before, I had assumed it was something with the Maven wrapper that made it do something without any goal(s). It's now very obvious how to just `compile` or `install` locally without worrying about checkstyle! thanks,

**Comment by jdimeo** [comment]
One more question to make sure I'm going in the right general direction. In the following example:
1) am I using the right verbiage? the right imperative form, etc.?
2) is it important to include `<name>` and `<type>` in the Javadocs? 
3) How does the .stg know what the sentinel/default value would be for methods like `get()` (so the Javadoc can say "if no value is present for the key, `0` is returned"? I'm seeing something in the unit test templates like `<(literal.(type1))>` which I think gets me close

I know this would be hashed out (pun intended) in the PR, but I want to get ahead of systematic changes if I'm off. thanks!

```
/**
 * Return whether or not this map contains the value.
 * @param value the value to test
 * @return if this collection contains the value
 */
boolean containsValue(<type> value);

/**
 * Iterate through each value in this map.
 * @param procedure the procedure to invoke for each value in this map.
 */
void forEachValue(<name>Procedure procedure);

/**
 * Return the values in this map as a separate collection. The returned collection is backed by the map, so any
 * changes made to the returned collection will affect the state of this map.
 * @return the values as a collection backed by this map
 */
Mutable<name>Collection values();
```

**Comment by nikhilnanivadekar** [comment]
@donraab can you opine?

**Comment by donraab** [comment]
Based on existing examples. 

Return should be Returns.
Iterate should be Iterates. 
etc.

The way I instinctively read this is:

"<containsValue> Returns whether or not this map contains the value." 
"<forEachValue> Iterates through each value in this map." 

Otherwise looks great.

**Comment by motlin** [comment]
The maven-wrapper is being integrated into maven and will be released together in the next version of maven, 3.7.0. So this information is temporary.

You can install the wrapper using `mvn -N io.takari:maven:0.7.7:wrapper`

It is normal to commit the wrapper to git, so that you can git clone and then immediately run `./mvwn`. I feel like we talked about this already though and I'm forgetting something. Why didn't we do this? Worry over how to deal with a second OSS license? At the time, I believed you had to commit .mvn/wrapper/ but you only have to commit mvnw and mvnw.cmd.

**Comment by jdimeo** [comment]
Initial work is complete on the `map` APIs over at https://github.com/eclipse/eclipse-collections/pull/887
